### DISCUSSION TITLE
Any way to easily expose HTTP routes

### DISCUSSION THREAD

**Discussion by ScreamZ** [next-safe-action/next-safe-action#128] [Q&A] [answered]
Hello,

I would like to rely fully on this library, but for now I'm using tRPC with such pattern : https://github.com/t3dotgg/server-actions-trpc-examples/tree/main/src/app

Resulting in duplicated validation for fields.

I would like to keep control over my data and be able to query it using non-js client.

Is there any way to easily support API endpoints using this? Can i just call my actions in express endpoints ?

**Comment by TheEdoRan** [comment]
Server Actions are normal functions, so you should be able to call them inside API routes/Express endpoints.

**Reply by ScreamZ** [reply, marked as answer]
I'll give it a try, this way I could get rid of tRPC for now. Will update this post

> [!NOTE]
> This is simple as using server action in request handlers of routes.
